apps/web/ee/lib/impersonation/ImpersonationProvider.ts

```diff
@@ -0,0 +1,62 @@
+import CredentialsProvider from "next-auth/providers/credentials";
+import { getSession } from "next-auth/react";
+
+import prisma from "@lib/prisma";
+
+const ImpersonationProvider = CredentialsProvider({
+  id: "impersonation-auth",
+  name: "Impersonation",
+  type: "credentials",
+  credentials: {
+    username: { label: "Username", type: "text " },
+  },
+  async authorize(creds, req) {
+    // @ts-ignore need to figure out how to correctly type this
+    const session = await getSession({ req });
+    if (session?.user.role != "ADMIN") {
+      throw new Error("You do not have permission to do this.");
+    }
+
+    if (session?.user.username == creds?.username) {
+      throw new Error("You cannot impersonate yourself.");
+    }
+
+    const user = await prisma.user.findUnique({
+      where: {
+        username: creds?.username,
+      },
+    });
+
+    if (!user) {
+      throw new Error("This user does not exist");
+    }
+
+    // Log impersonations for audit purposes
+    await prisma.impersonations.create({
+      data: {
+        impersonatedBy: {
+          connect: {
+            id: session.user.id,
+          },
+        },
+        impersonatedUser: {
+          connect: {
+            id: user.id,
+          },
+        },
+      },
+    });
```

Comment:
Good thinking on logging the impersonations, maybe in the future we could also add some metadata about actions done during impersionation to improve the -audit- aspects, so if I were to impersonate and change X, y, z, all those actions where to be logged into the impersionating session?

Surely not for this PR/MVP, but just thought for the future

Reply:
Im pretty sure this was the plan :)